### PR TITLE
feat: add the envelop object annotation to support extra work for enveloped objects

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

add the envelop object annotation to the master resource snapshot to support extra work for enveloped objects.

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

UT/E2E

### CODE REVIEW

File: pkg/controllers/clusterresourceplacement/controller_test.go [hunk 1]
```diff
@@ -1688,24 +1696,6 @@ func TestGetOrCreateClusterResourceSnapshot_failure(t *testing.T) {
 				},
 			},
 		},
-		{
-			// Should never hit this case unless there is a bug in the controller or customers manually modify the clusterResourceSnapshot.
```
Comment (zhiying-lin) [inline pkg/controllers/clusterresourceplacement/controller_test.go:1692]: may i know why we remove this test case?

File: pkg/utils/annotations/annotations_test.go [hunk 2]
```diff
@@ -305,3 +304,82 @@ func TestExtractNumberOfResourceSnapshots(t *testing.T) {
...
+			},
+			wantError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := ExtractNumberOfEnvelopeObjFromResourceSnapshot(tc.snapshot)
+			if gotErr := err != nil; gotErr != tc.wantError {
+				t.Fatalf("ExtractNumberOfEnvelopeObjFromResourceSnapshot teset `%s` got err %v, want err %v", tc.name, err, tc.wantError)
```
Comment (zhiying-lin) [inline pkg/utils/annotations/annotations_test.go:378]:
```suggestion
				t.Fatalf("ExtractNumberOfEnvelopeObjFromResourceSnapshot() got err %v, want err %v", tc.name, err, tc.wantError)
```

The tc.name will be always printed as we define it in the t.Run(t.name, xxx)

File: pkg/utils/annotations/annotations_test.go [hunk 3]
```diff
@@ -305,3 +304,82 @@ func TestExtractNumberOfResourceSnapshots(t *testing.T) {
...
+			}
+			if !tc.wantError && got != tc.want {
+				t.Fatalf("ExtractNumberOfEnvelopeObjFromResourceSnapshot teset `%s` = %v, want err %v", tc.name, got, tc.want)
```
Comment (zhiying-lin) [inline pkg/utils/annotations/annotations_test.go:381]:
```suggestion
				t.Fatalf("ExtractNumberOfEnvelopeObjFromResourceSnapshot() `%s` = %v, want err %v", tc.name, got, tc.want)
```

File: pkg/controllers/clusterresourceplacement/controller_test.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/controllers/clusterresourceplacement/controller_test.go:1692]: Is this not the same as the one above?

Comment (zhiying-lin) [reply pkg/controllers/clusterresourceplacement/controller_test.go:1692]: oh, you're right, it's duplicated.

File: test/e2e/enveloped_object_placement_test.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+					return fmt.Errorf("CRP status diff (-got, +want): %s", diff)
+				}
+				if !strings.Contains(placementStatus.FailedResourcePlacements[0].Condition.Message, "unsupported scope") {
+					return fmt.Errorf("CRP failed resource placement does not have unsupported scope message")
+				}
+				if diff := cmp.Diff(placementStatus.Conditions, resourcePlacementApplyFailedConditions(crp.Generation), crpStatusCmpOptions...); diff != "" {
+					return fmt.Errorf("CRP status diff (-got, +want): %s", diff)
+				}
+			} else {
+				// the cluster is stuck behind a rollout schedule since we now have 1 cluster that is not in applied ready status
```
Comment (zhiying-lin) [inline test/e2e/enveloped_object_placement_test.go:260]: should we wait until these clusters are in the stable state then compare the status?

We may stop validating if hit this "else" condition and may miss the regression.

File: test/e2e/enveloped_object_placement_test.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+			if len(placementStatus.FailedResourcePlacements) != 0 {
+				applyFailed = true
+			}
+		}
+		if !applyFailed {
+			return fmt.Errorf("CRP status does not have failed placement")
+		}
+		for _, placementStatus := range crp.Status.PlacementStatuses {
+			// this is the cluster that got the new enveloped resource that was malformed
+			if len(placementStatus.FailedResourcePlacements) != 0 {
```
Comment (zhiying-lin) [inline test/e2e/enveloped_object_placement_test.go:249]: Once the state is stable, you can compare the whole CRP status similar to https://github.com/Azure/fleet/blob/main/test/e2e/placement_selecting_resources_test.go#L802.

The readability will be better.

Comment (ryanzhang-oss) [reply test/e2e/enveloped_object_placement_test.go:249]: there is no guarantee which cluster has the failed resource placement

File: test/e2e/setup_test.go [hunk 6]
```diff
@@ -47,7 +47,7 @@ const (
 )
 
 const (
-	eventuallyDuration   = time.Minute * 10
+	eventuallyDuration   = time.Minute * 3
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply test/e2e/setup_test.go:50]: 3 mins should be enough for all steps (all e2e passes)

File: test/e2e/setup.sh [hunk 7]
```diff
@@ -65,7 +65,8 @@ helm install hub-agent ../../charts/hub-agent/ \
     --set enableWebhook=true \
     --set webhookClientConnectionType=service \
     --set enableV1Alpha1APIs=false \
-    --set enableV1Beta1APIs=true
+    --set enableV1Beta1APIs=true \
+    --set enableGuardRail=false
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply test/e2e/setup.sh:69]: good catch, local environment slipped in

File: test/e2e/enveloped_object_placement_test.go (hunk 4 above)
Comment (ryanzhang-oss) [reply test/e2e/enveloped_object_placement_test.go:260]: this is the stable state. What else can happen?

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**